bodyswap.lua

```diff
@@ -122,22 +151,94 @@ function swapAdvUnit(newUnit)
     df.global.adventure.player_id = newNem.id
     df.global.world.units.adv_unit = newUnit
     oldUnit.idle_area:assign(oldUnit.pos)
+    local pos = xyz2pos(dfhack.units.getPosition(newUnit))
+    -- reveal the tiles around the bodyswapped unit
+    reveal_around(pos)
+    -- Focus on the revealed pos
+    dfhack.gui.revealInDwarfmodeMap(pos, true)
+end
 
-    dfhack.gui.revealInDwarfmodeMap(xyz2pos(dfhack.units.getPosition(newUnit)), true)
+-- shamelessly copy pasted from gui/sitemap.lua
+local function get_unit_choices()
+    local choices = {}
+    for _, unit in ipairs(df.global.world.units.active) do
+        if not dfhack.units.isActive(unit) or
+            dfhack.units.isHidden(unit)
+        then
+            goto continue
+        end
+        local name = dfhack.units.getReadableName(unit)
+        table.insert(choices, {
+            text=name,
+            unit=unit,
+            search_key=dfhack.toSearchNormalized(name),
+        })
+        ::continue::
+    end
+    return choices
+end
+
+local function swapAdvUnitPrompt()
+    local choices = get_unit_choices()
+    dialogs.showListPrompt('bodyswap', "Select a unit to bodyswap to:", COLOR_WHITE,
+        choices, function(id, choice)
+            swapAdvUnit(choice.unit)
+        end, nil, nil, true)
+end
+
+function getHistoricalSlayer(unit)
+    local histFig = unit.hist_figure_id ~= -1 and df.historical_figure.find(unit.hist_figure_id)
+    if not histFig then
+        return
+    end
+
+    local deathEvents = df.global.world.history.events_death
+    for i = #deathEvents - 1, 0, -1 do
+        local event = deathEvents[i] --as:df.history_event_hist_figure_diedst
+        if event.victim_hf == unit.hist_figure_id then
+            return df.historical_figure.find(event.slayer_hf)
+        end
+    end
+end
+
+if positionals[1] == 'linger' then
+    local adventurer = dfhack.world.getAdventurer()
+    if not adventurer.flags2.killed then
+        qerror("Your adventurer hasn't died yet!")
+    end
+
+    local slayerHistFig = getHistoricalSlayer(adventurer)
+    local slayer = slayerHistFig and df.unit.find(slayerHistFig.unit_id)
+    if not slayer then
+        slayer = df.unit.find(adventurer.relationship_ids.LastAttacker)
+    end
+    if not slayer then
+        qerror("Killer not found!")
+    elseif slayer.flags2.killed then
+        local slayerName = ""
+        if slayer.name.has_name then
+            slayerName = ", " .. dfhack.TranslateName(slayer.name) .. ","
+        end
+        qerror("Your slayer" .. slayerName .. " is dead!")
+    end
+
+    swapAdvUnit(slayer)
+    return
 end
 
 if not dfhack_flags.module then
     if df.global.gamemode ~= df.game_mode.ADVENTURE then
         qerror("This script can only be used in adventure mode!")
     end
 
-    local unit = args.unit and df.unit.find(tonumber(args.unit)) or dfhack.gui.getSelectedUnit()
+    local unit = options.unit ~= -1 and df.unit.find(options.unit) or dfhack.gui.getSelectedUnit()
```

Comment:
```suggestion
    local unit = options.unit == -1 and dfhack.gui.getSelectedUnit(true) or df.unit.find(options.unit)
```

Reply:
to avoid the case of a bad unit id being specified and the selection incorrectly falling through to a selected unit

also passed `true` to `getSelectedUnit` to it doesn't print out an error when no unit is selected